### PR TITLE
[flutter_tools] refactor Chrome launch logic to remove globals/statics

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Refactors the Chrome launch logic to associate a `Chromium` instance with a specific  `ChromiumLauncher` instead of using statics on the old `ChromeLauncher`. Removes globals from these classes, along with the web devices classes.

This change will eventually allow more browser instances, and as an example I modified the chrome code to support different chromium based browsers. Included is some rudimentary support for Chromium-based edge, which is tested but not exposed anywhere. Exposing this would require a change in dwds: https://github.com/dart-lang/webdev/issues/962, but otherwise it works locally - modulo some follow up work to determine the version and executable locations on Linux and macOS.

Work towards https://github.com/flutter/flutter/issues/55322

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Bug: cannot currently close browser correctly

File: packages/flutter_tools/lib/src/commands/devices.dart [hunk 1]
```diff
@@ -60,7 +58,9 @@ class DevicesCommand extends FlutterCommand {
 
     final List<Device> devices = await deviceManager.refreshAllConnectedDevices(timeout: timeout);
 
-    if (devices.isEmpty) {
+    if (boolArg('machine')) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/devices.dart:61]: Empty list was printing non-json: fixed the format and added a test

File: packages/flutter_tools/lib/src/web/web_device.dart [hunk 2]
```diff
@@ -177,27 +157,139 @@ class ChromeDevice extends Device {
...
+        version = result.stdout as String;
+      }
+    }
+    return version.trim();
+  }
+
+}
+
+/// The Microsoft Edge browser based on Chromium.
+class MicrosoftEdgeDevice extends ChromiumDevice {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/web/web_device.dart:219]: The edge device isn't hooked up yet

File: packages/flutter_tools/lib/src/web/web_device.dart [hunk 3]
```diff
@@ -244,7 +341,7 @@ class WebServerDevice extends Device {
   Future<bool> get isLocalEmulator async => false;
 
   @override
-  bool isSupported() => featureFlags.isWebEnabled;
+  bool isSupported() => true;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/web/web_device.dart:344]: If featureFlags.isWebEnabled == false, then the device discoverer isn't supported which is redundant with this check

File: packages/flutter_tools/lib/src/web/web_validator.dart [hunk 4]
```diff
@@ -54,3 +46,41 @@ class WebValidator extends DoctorValidator {
...
+
+  @override
+  final ChromiumLauncher _chromiumLauncher;
+
+  @override
+  String get _name => 'Chrome';
+}
+
+/// A validator that checks whethere Edge is installed and can run.
+class EdgeValidator extends ChromiumValidator {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/web/web_validator.dart:70]: Not used yet

File: packages/flutter_tools/lib/src/web/web_device.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/web/web_device.dart:219]: Should we note somewhere in the code that this isn't hooked up yet?

File: packages/flutter_tools/lib/src/web/web_device.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/web/web_device.dart:344]: I understand that currently this check is redundant. Could a future refactor change that however?

**Review by christopherfujino — APPROVED** [review]
LGTM, with a few minor questions. One more question: are we tracking support for other browsers on an issue (or somewhere else)?

**Comment by jonahwilliams** [comment]
I filled issues for Edge and Safari:

https://github.com/flutter/flutter/issues/55322
https://github.com/flutter/flutter/issues/55323

File: packages/flutter_tools/lib/src/web/web_device.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/web/web_device.dart:220]: Done, here and in the validator

File: packages/flutter_tools/lib/src/web/web_device.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/web/web_device.dart:345]: There is an existing workflow test that confirms that the web devices aren't included if the config is disabled: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/web/workflow_test.dart#L57

I added another test case to the `web/devices_test` that verifies no devices are returned if the config is disabled

**Pull request merged**